services/storage/src/simcore_service_storage/access_layer.py

```diff
@@ -0,0 +1,266 @@
+
+    ## GROUP-BASED METHOD:
+        The second method is designed to dynamically share a resource among groups of users. A group
+        defines a set of rules that apply to a resource and users can be added to the group dynamically.
+        So far, there are two resources that define access rights (AR):
+            - one applies to projects (read/write/delete) and
+            - the other to services (execute/write)
+        The project access rights are set in the column "access_rights" of the "projects" table .
+        The service access rights has its own table: service_access_rights
+
+        Access rights apply hierarchically, meaning that the access granted to a project applies
+        to all nodes inside and stored data in nodes.
+
+        How do these two AR coexist?: Access to read, write or delete a project are defined in the project AR but execution
+        will depend on the service AR attached to nodes inside.
+
+        What about stored data?
+        - data generated in nodes inherits the AR from the associated project
+        - data generated in API uses full AR provided by ownership (i.e. user_id in files_meta_data table)
+
+"""
+
+
+import logging
+from dataclasses import dataclass
+from typing import Dict, List, Optional
+from uuid import UUID
+
+import sqlalchemy as sa
+from aiopg.sa.connection import SAConnection
+from aiopg.sa.result import ResultProxy, RowProxy
+from simcore_postgres_database.storage_models import file_meta_data, user_to_groups
+from sqlalchemy.sql import text
+
+logger = logging.getLogger(__name__)
+
+
+ProjectID = str
+
+
+@dataclass
+class AccessRights:
+    read: bool
+    write: bool
+    delete: bool
+
+    @classmethod
+    def all(cls) -> "AccessRights":
+        return cls(True, True, True)
+
+    @classmethod
+    def none(cls) -> "AccessRights":
+        return cls(False, False, False)
+
+
+class NotAllowedError(Exception):
+    ...
+
+
+async def _get_user_groups_ids(conn: SAConnection, user_id: int) -> List[int]:
+    stmt = sa.select([user_to_groups.c.gid]).where(user_to_groups.c.uid == user_id)
+    rows = await (await conn.execute(stmt)).fetchall()
+    user_group_ids = [g.gid for g in rows]
+    return user_group_ids
+
+
+def _aggregate_access_rights(
+    access_rights: Dict[str, Dict], group_ids: List[int]
+) -> AccessRights:
+    try:
+        prj_access = {"read": False, "write": False, "delete": False}
+        for gid, grp_access in access_rights.items():
+            if int(gid) in group_ids:
+                for operation in grp_access:
+                    prj_access[operation] |= grp_access[operation]
+
+        return AccessRights(**prj_access)
+    except KeyError:
+        # NOTE: database does NOT include schema for json access_rights column!
+        logger.warning(
+            "Invalid entry in projects.access_rights. Revoking all rights [%s]",
+            access_rights,
```

Comment:
yes the plan is to move the json column to a separate table (similar as with the services_access_rights) it would allow relations between tables that the json columns do not allow.